### PR TITLE
Fixing expression compiler support for multi-dimensional arrays of rank 1

### PR DESCRIPTION
**Pull request by bartdesmet**
This fixes issue #4161. The following issues were found and resolved for multi-dimensional arrays of rank 1 (MDR1s):

- ArrayLength nodes could be created over MDR1s, leading to a `VerificationException` upon executing the `ldlen` instruction. Creation of those nodes is now rejected, just like it was for any other multi-dimensional array.
- ArrayAccess nodes could be used to read from and write to MDR1s, leading to `VerificationException` for the `ldelem` and `stelem` instructions. Those compile to `Get` and `Set` method calls just fine now.
- ArrayIndex nodes could be used to read from MDR1s, leading to `VerificationException` for the `ldelem` instruction. Those compile to `Get` method calls just fine now.

Expression interpreter works fine thanks to the use of `GetValue` and `SetValue` methods for late-bound access to the arrays.

### CODE REVIEW

File: src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs [hunk 1]
```diff
@@ -730,5 +730,13 @@ internal static bool IsFloatingPoint(Type type)
             }
         }
 #endif 
+
+        public static bool IsVector(this Type type)
+        {
+            // Unfortunately, the IsSzArray property of System.Type is inaccessible to us,
+            // but the ToString representation of the type is sufficient to differentiate
+            // a vector (X[]) from a multi-dimensional array of rank 1 (X[*]).
+            return type.GetArrayRank() == 1 && type.ToString().EndsWith("[]");
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs]: See https://github.com/dotnet/coreclr/issues/1877 for a CoreCLR issue filed to improve on the ToString workaround.

Comment (JonHanna) [inline src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs]: Would checking that the lower bound is zero be worth doing? I imagine it's faster than string comparison (but haven't checked), but then it's probably optimising the less-common case, so perhaps not worth it.

Comment (mikedn) [inline src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs]: What about 
```C#
return type == type.GetElementType().MakeArrayType();
```

Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs]: Great idea. Will revisit the PR accordingly.

**Comment by stephentoub** [comment]
cc: @VSadov

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.Expressions.cs [hunk 2]
```diff
@@ -311,15 +319,23 @@ private void EmitSetIndexCall(IndexExpression node, Type objectType)
...
             {
                 // Multidimensional arrays, call set
-                _ilg.Emit(OpCodes.Call, node.Object.Type.GetMethod("Set", BindingFlags.Public | BindingFlags.Instance));
+                _ilg.Emit(OpCodes.Call, arrayType.GetMethod("Set", BindingFlags.Public | BindingFlags.Instance));
             }
             else
             {
                 // For one dimensional arrays, emit store
-                _ilg.EmitStoreElement(node.Type);
+                _ilg.EmitStoreElement(arrayType.GetElementType());
```
Comment (stephentoub) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.Expressions.cs]: Are these comments still correct?  Seems like it should be multidimensional arrays (including one-dim arrays with non-zero lower-bounds) vs vectors?

File: src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs [hunk 3]
```diff
@@ -730,5 +730,12 @@ internal static bool IsFloatingPoint(Type type)
             }
         }
 #endif 
+
+        public static bool IsVector(this Type type)
+        {
+            // Unfortunately, the IsSzArray property of System.Type is inaccessible to us,
+            // so we use a little equality comparison trick instead:
+            return type == type.GetElementType().MakeArrayType();
```
Comment (stephentoub) [inline src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs]: If the object references are equal, the types are the same.  But is it guaranteed that if they're not equal they're not the same type?  I know the runtime caches Type objects, but I'm not sure what the caching policy is, whether it's based on weak references (in which case this should be fine) or something else (in which case maybe there's a problem?)

Comment (stephentoub) [inline src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs]: Oh, nevermind, Type implements the equality operator.

**Comment by stephentoub** [comment]
LGTM

**Comment by VSadov** [comment]
LGTM

**Pull request merged**